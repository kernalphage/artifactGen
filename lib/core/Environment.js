import {_} from "lodash";
export class Environment{
    constructor(id){
        this.defs = {};
        this.id = id;
    }
    applyDefinition(target, value) {
        // TODO: split some of this into a "findTarget" function
        if(!(target instanceof Array)){
            target = target.split(".");
        }
        let cur = this;
        _.forEach(target.slice(0,-1), (loc) => {
            if(!(loc in cur.defs)){
                cur.defs[loc] = new Environment();   // theres a lot of things this node could be
            }
            cur = cur.defs[loc];
        });
        let ass = _.last(target);
        cur.defs[ass] = value;
    }
    get(target) {
        if(!(target instanceof Array)){
            target = target.split(".");
        }

        let res = _.reduce(target, (cur, loc)=>{
            if(cur instanceof Environment) {
                return cur.defs[loc];
            } else if(cur.value) {
                return cur.value.defs[loc];
            }
        }, this) || [];
    // TODO: this is the wrong place to do this. Time to get automagic with types
        if(res.length == 1){
            return res[0];
        }
        return res;
    }

    // TODO: Re-read the chapter on environments. This is definitely not groovy
    export(){
        let output =  _.map(this.defs, (value, key)=>{
            let out = key+":";
            out+=findString(value);
            return out;
        }).join("\n");
        return output;
    }
}

function findString(o){
    if(o instanceof Environment){
        return "\n" + o.export();
    } else if (o instanceof Array){
        return _.flatMap(o, findString).join(" ");
    } else if(o.value){
        return findString(o.value);
    } else {
        return "TOSTRING" + o.toString();
    }
};