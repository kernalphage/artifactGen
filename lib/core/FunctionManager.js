import {_} from "lodash";

let _functions = {};

function register(fnName, fn, hints){
    _functions[fnName] = [fn, hints];
}

function applyFN(fnName, context, args){
    return _functions[fnName][0](context, ...args);
}

function getHint(fnName){
    return _functions[fnName][1];
}

function choice(context, ...values){
    return context.visit(_.sample(values));
}
register("c", choice);
register("choice", choice);

register("chance", (context, chance, expr)=>{
    if(Math.random() * 100 > context.visit(chance)){
        return context.visit(expr);
    } 
    return null;
}, ["Chance:NumberExpression", "expr:Expression"]);

register("many", (context, num, expr)=>{
    let times = context.visit(num);
    let res = _.times(times, ()=>{
        return context.visit(expr);
    });
    console.log("Applying function " + times + " with many and got result " + JSON.stringify(res));
    return res;
}, ["times:NumberExpression", "expr:Expression"]);

register("if", (predicate, ifCase, thenCase)=>{
    if(context.visit(predicate)){
        return context.visit(ifCase);
    }
    else {
        return context.visit(thenCase);
    }
}, ["predicate:Value", "ifCase:Expression", "thenCase:Expression"]);

module.exports.register = register;
module.exports.applyFN = applyFN;
module.exports.getHint = getHint;