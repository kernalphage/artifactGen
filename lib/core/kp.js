import {
    _
} from 'lodash';

// TODO: Clean this up with something smarter
// How does the saying go? 
// Javascript pursues other languages down alleyways to beat them unconscious and rifle their pockets for new syntax
// This is an attempt to implement typeclasses and some metarpogramming shenenagians using 
// A mixture of Symbol(), matching dictionaries, and convienent ducktyping 
// Transforms: TypeData: {cat:['tail', "teeth"], bird:['wing']} => [ 
//  {'cat':Symbol('cat'), 'bird':Symbol('bird')}      // SymbolAccessor =_.mapValues(TypeData, (v,k)=>{return Symbol(k);});
//  {Symbol('cat'):['tail'], Symbol('bird'):['wing']} // PropertyAccessor = _.mapKeys(TypeData, (v,k)=>{return Ex[k];});
// ]

// SymbolAccessor: string => symbol, 
// PropertyAccessor: symbol => properties

function TypeFactory(Typeclass, ClassName) {
    // Creates an expression of type symbol, and populates it if possible
    return function instantiate(symbol, ...args) {
        if (!symbol) {
            throw new Error("Symbol does not exist!");
        }
        if (!(_.has(Typeclass, symbol))) {
            throw new Error("Symbol " + symbol.toString() + " Not in category " + ClassName);
        }
        let obj = _.zipObject(Typeclass[symbol], args);
        obj.Type = symbol;
        obj.Klass = ClassName; // Not sure if this is needed but hey, why not
        return obj;
    };
}

export function FindFromSymbol(SymbolAccessor, symbol) {
    return _.findKey(SymbolAccessor, (sym) => {
        return sym == symbol;
    });
}

export function MakeTypeclass(TypeData, ClassName) {
    let SymbolAccessor = _.mapValues(TypeData, (v, k) => {
        return Symbol(k);
    });
    let PropertyAccessor = _.mapKeys(TypeData, (v, k) => {
        return SymbolAccessor[k];
    });
    return [TypeFactory(PropertyAccessor, ClassName), SymbolAccessor];
}

export function isSameType(a, b) {
    a = (a instanceof Object) ? a.Type : a;
    b = (b instanceof Object) ? b.Type : b;
    return a === b; 
}

// Different useful functions

// [a] => [Symbol(a)] 
export function makeEnum(arr) {
    let obj = {};
    for (let val of arr) {
        obj[val] = Symbol(val);
    }
    return Object.freeze(obj);
}

export function isDigit(c) {
    return c >= '0' && c <= '9';
}
export function isAlpha(c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}
export function isAlphaNumeric(c) {
    return isAlpha(c) || isDigit(c);
}
export function isEmoji(c){
    let filter = /[\u{1f300}-\u{1f5ff}\u{1f900}-\u{1f9ff}\u{1f600}-\u{1f64f}\u{1f680}-\u{1f6ff}\u{2600}-\u{26ff}\u{2700}-\u{27bf}\u{1f1e6}-\u{1f1ff}\u{1f191}-\u{1f251}\u{1f004}\u{1f0cf}\u{1f170}-\u{1f171}\u{1f17e}-\u{1f17f}\u{1f18e}\u{3030}\u{2b50}\u{2b55}\u{2934}-\u{2935}\u{2b05}-\u{2b07}\u{2b1b}-\u{2b1c}\u{3297}\u{3299}\u{303d}\u{00a9}\u{00ae}\u{2122}\u{23f3}\u{24c2}\u{23e9}-\u{23ef}\u{25b6}\u{23f8}-\u{23fa}]/ug;
    return c.match(filter);
}
export function isLiteral(c){
    return isAlphaNumeric(c) || isEmoji(c);
}

export function randomRangeInt(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min)) + min; //The maximum is exclusive and the minimum is inclusive
  }


  // Support for unicode
export function getWholeChar(str, i) {
    var code = str.charCodeAt(i);

    if (Number.isNaN(code)) {
        return ''; // Position not found
    }
    if (code < 0xD800 || code > 0xDFFF) {
        return str.charAt(i);
    }

    // High surrogate (could change last hex to 0xDB7F to treat high private
    // surrogates as single characters)
    if (0xD800 <= code && code <= 0xDBFF) {
        if (str.length <= (i + 1)) {
            throw 'High surrogate without following low surrogate';
        }
        var next = str.charCodeAt(i + 1);
        if (0xDC00 > next || next > 0xDFFF) {
            throw 'High surrogate without following low surrogate';
        }
        return str.charAt(i) + str.charAt(i + 1);
    }
    // Low surrogate (0xDC00 <= code && code <= 0xDFFF)
    if (i === 0) {
        throw 'Low surrogate without preceding high surrogate';
    }
    var prev = str.charCodeAt(i - 1);

    // (could change last hex to 0xDB7F to treat high private
    // surrogates as single characters)
    if (0xD800 > prev || prev > 0xDBFF) {
        throw 'Low surrogate without preceding high surrogate';
    }
    // We can pass over low surrogates now as the second component
    // in a pair which we have already processed
    return false;
}