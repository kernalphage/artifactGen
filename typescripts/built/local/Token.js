"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Token {
    constructor(symbol, string, line, value = null) {
        this.symbol = symbol;
        this.string = string;
        this.line = line;
        this.value = (value == null) ? string : value;
    }
    toString() {
        return ">" + this.string + "< (line " + this.line + ")";
    }
}
exports.Token = Token;
