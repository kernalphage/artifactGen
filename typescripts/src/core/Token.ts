export const enum tk {
    AT,
    BANG,
    BAR,
    COLON,
    COMMA,
    DOLLAR,
    DOT,
    EOF,
    EQUALS,
    ERROR,
    HASH,
    LEFT_BRACKET,
    LEFT_CURLY,
    LEFT_PAREN,
    LITERAL,
    MINUS,
    NEWLINE,
    NUMBER,
    PLUS,
    QUESTION,
    RIGHT_BRACKET,
    RIGHT_CURLY,
    RIGHT_PAREN,
    SEMICOLON,
    SLASH,
    STAR,
}

export class Token {
    readonly value: any;
    constructor(readonly symbol: tk, readonly string: string, readonly line: Number, value: any = null) {
        this.value = (value == null) ? string : value;
    }

    toString() {
        return ">" + this.string + "< (line " + this.line + ")";
    }
}