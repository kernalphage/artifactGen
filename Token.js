import {makeEnum} from "./kp.js";
// maybe makeEnum should return a function, making all my calls tk("LEFT PAREN"); 
// i think the extra quotes are worth some type safety
export const tk = makeEnum([
    "LEFT_PAREN",
    "RIGHT_PAREN",
    "LEFT_BRACKET",
    "RIGHT_BRACKET",
    "LEFT_CURLY",
    "RIGHT_CURLY",
    "COMMA",
    "DOT",
    "MINUS",
    "PLUS",
    "SEMICOLON",
    "COLON",
    "SLASH",
    "STAR",
    "BANG",
    "AT",
    "DOLLAR",
    "HASH",
    "QUESTION",
    "EQUALS",

    "COMMENT",
    "LITERAL",
    "NUMBER",
    "EOF", 
    "ERROR",
    "BAR",
    
]);
    


export class Token{
    constructor(symbol, string, line, value){
        this.symbol = symbol;
        this.string = string;
        this.value = (value == null) ? string : value;
        this.line = line;
    }    

    toString(){
        let va = " with value " + this.value;
        return this.symbol.toString() + ">" + this.string + "<" + va;
    }
};