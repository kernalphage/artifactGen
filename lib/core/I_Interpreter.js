
import { FindFromSymbol } from './kp.js';
import {Token} from './Token.js';
import {_} from "lodash";
import {pr} from "./Parser.js";

export class I_Interpreter {
    constructor(){
        this.contexts = [];
    }
    push_context(env) {
        this.contexts.push(env);
    }
    peek_context() {
        return _.last(this.contexts);
    }
    pop_context() {
        return this.contexts.pop();
    }
    execute(definitions){
        return this.visitMany(definitions);
    }

    // TODO: i don't like this visit vs visit_many
    // i think I had a binding issue 
    visitMany(exprs){
        let ctx = this.peek_context();
        var ret = [];
        if(exprs instanceof Array){
            for(var i=0; i < exprs.length; i++){
                ret.push(this.visit(exprs[i], ctx));
            }
            return ret;
        }
        return [this.visit(exprs, ctx)];
    }
    // TODO: preVisit() and postVisit() to push/pop state?
    visit(expr){
        if(expr instanceof Token){
            return expr.value;
        }
        let key = FindFromSymbol(pr, expr.Type);
        let fn = this.visitors[key];
        if(fn){
            // TODO: I could pass in recursion depth, entire tree state, lots of fun stuff
            return fn.apply(this, [expr, this.peek_context()]);
        } else {
            console.log("Could not find expression " + expr.Type.toString());
            return JSON.stringify(expr);
            //throw this.RuntimeError("Visitor function for Expresion of type " + expr.Type.toString() + " does not exist");
        } 
    }
}