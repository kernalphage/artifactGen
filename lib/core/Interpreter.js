import {I_Interpreter} from "./I_Interpreter.js";
import {tk } from "./Token.js";
import { randomRangeInt } from './kp.js';
import {_} from "lodash";
import {Environment} from "./Environment.js";
import { kMaxLength } from "buffer";

export class BasicInterpreter extends I_Interpreter {
    execute(expressions){
        super.execute(expressions);
        this.treeShake();
        return this.definitions;
    }
    push_reference(loc){
        let ref = {loc: loc, value: null, tries: 0, toString:()=>{return "ASDF" + this.value;}};
        // TODO: cache references 
        this.toResolve.push(ref);
        return ref;
    }

    // TODO: Actual dependency resolution
    treeShake(){
        let open = this.toResolve;

        let failsafe = 100;
        while(open.length > 0 && (failsafe-->0)){
            let cur = open.pop();
            let val = this.findRvalue(cur.loc);
            if(val){
                cur.value = val;
            } else {
                cur.tries ++;
                open.push(cur);
                // tODO: shuffle before pushing? 
            }
        }
    }

    findRvalue(val){
        try{
            let cur = this.definitions.get(val);
            return cur;
        } catch(e) {
            console.log("Error finding rvalue " + e);
            return null;
        }
    }

    export(){
        return this.definitions.export();
    }
    constructor(){
        super();
        this.definitions = new Environment();
        this.toResolve = [];

        this.visitors = {
            Assignment:  (expr, ctx)=>{
                let chosenAssignment = _.sample(expr.values);
                let out = _.zip(expr.targets, chosenAssignment.statements).map((assignment) => {
                    let [target, statements] = assignment;
                    target = target && this.visit(target);
                    console.log("target is " + JSON.stringify(target));
                    statements = statements && this.visitMany(statements);
                    let result = _.sample(statements);
                    console.log(JSON.stringify(result));
                    console.log("statement is " + result);
                    ctx.applyDefinition(target, result);
                });
                return out;
            },
            Base: (expr)=>{                
                return this.visitMany(expr.values);
            },
            Definition:  (expr)=>{
                let name = expr.id.string;
                let def = new Environment(name);
                this.push_context(def);
                this.visitMany(expr.assignments);
                this.pop_context();
                this.definitions.applyDefinition(name, def);
                return def;
            },
            // function
            Locator:  (expr)=>{
               return _.map(expr.location, 'value');
            },
            LValue:  (expr)=>{
                // TODO: Split on relative vs absolute
                return this.visit(expr.locator);
            },
            Number:  (expr)=>{
                if(expr.values.length == 1){
                    return expr.values[0].value;
                } else {
                    // TODO: Support floats
                    return randomRangeInt(... _.map(expr.values, 'value'));
                }
            },
            RValue:  (expr, ctx)=>{
                let loc = this.visit(expr.locator);
                try {
                    let rval = null;
                    if(expr.reftype.symbol == tk.HASH ){
                        rval = this.definitions.get(loc);
                    } else {
                        rval = ctx.get(loc);
                    }
                    return rval; 
                } finally{
                    let rval = null;
                    if (expr.reftype.symbol == tk.HASH) {
                        rval = loc;
                    } else {
                        rval = _.concat(ctx.id, loc)
                    }
                    return this.push_reference(rval); 
                }
            },
            // sideEffect
            Statement: (expr)=>{
                return this.visitMany(expr.statements);
            },
            ExFunction: (expr)=>{
                let loc = this.visit(expr.locator);
                let params = this.visitMany(expr.parameters);
                return loc + "(" + params.join(", ") + ")";
            }
        };
    }
}