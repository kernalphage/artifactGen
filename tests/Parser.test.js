import ArtifactGenerator from "..";
let {Parser, Scanner} = ArtifactGenerator.Core;
// let Scanner = ArtifactGenerator.Core;
import{_} from 'lodash';
import { tk } from "../lib/core/Token";

test('basic definition', () => {
  let source = `
 [item]
 pi = 3.14;
 randomItem = 1:10;
 metalColor, metalValue  = gold, 200 | silver, 100 | copper, 10; 
 description = this is a @metalColor ring that is worth @metalValue 

 [Second]
 math = #maths;
 value = "special 324";
 function = !fun(ction);

 [maths]
 pi = 3.14;
 alsoPi = @pi;
 e = 3
`;
  let scan = new Scanner(source);
  let parse = new Parser(scan.tokens);
  expect(parse.definitions.length).toBe(3);
});

test('find_some success', () => {

  let tBase     = [new Scanner("a, b, c, d"),   tk.LITERAL, tk.COMMA];
  let tTrail    = [new Scanner("a# b# c# d |"), tk.LITERAL, tk.HASH];
  let tTrail2   = [new Scanner("a, b, c, d,|"), tk.LITERAL, tk.COMMA, tk.BAR];

  let shouldSucceed = [tBase, tTrail, tTrail2];
  let parse     = new Parser([]);
  shouldSucceed.forEach(element => {
    let [tokens, ...params] = element;
    parse.reset(tokens.tokens);
    expect(parse.find_some(...params)).toBeTruthy();
  });
  parse.reset()
});

test('find_some failures', () => {

  let tTrail3 = [new Scanner("a, b, c, d,|"), tk.LITERAL, tk.COMMA];
  let tTrail4 = [new Scanner("a, b, c, d,"), tk.LITERAL, tk.COMMA];
  let tNone = [new Scanner("1"), tk.LITERAL, tk.COMMA];
  let tNone2 = [new Scanner("1,2,3"), tk.LITERAL, tk.COMMA];

  let shouldFail = [tTrail3, tTrail4, tNone, tNone2];
  let parse = new Parser([]);
  shouldFail.forEach(element => {
    let [tokens, ...params] = element;
    parse.reset(tokens.tokens);
    expect(() => {
      parse.find_some(...params)
    }).toThrow(/Parser error/);
  });
  parse.reset()
});

// TODO: need more code covereage here
test('basic parsing mixups', () => {
  let sources = [`
  [noSemicolon]
two_lines = that will @break 
without = semicolons
` , `
[noEquals]
error here; 
can you believe it
`];
let errors = [];
  sources.forEach((source)=>{
    let scan = new Scanner(source);
    let parse = new Parser(scan.tokens);
    expect(parse.definitions.length).toBe(0);
    errors.push(parse.errors);
  });
  expect(errors).toMatchSnapshot();

});
