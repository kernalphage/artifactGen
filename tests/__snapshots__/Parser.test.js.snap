// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic parsing mixups 1`] = `
Array [
  Array [
    [ParserError: Parser error at >=< (line 4): Found an unexpected character in base expression (are you missing a semicolon?) ],
  ],
  Array [
    [ParserError: Parser error at >here< (line 3): Missing '=' in assignment definition(Expected [Symbol(EQUALS)])],
  ],
]
`;
